lib/ipam/ipam.go

```diff
@@ -621,18 +838,38 @@ func (c ipamClient) RemoveIPAMHost(ctx context.Context, host string) error {
 	}
 
 	// Release affinities for this host.
-	c.ReleaseHostAffinities(ctx, hostname)
+	if err := c.ReleaseHostAffinities(ctx, hostname); err != nil {
+		log.WithError(err).Errorf("Failed to release IPAM affinities for host")
+		return err
+	}
 
-	// Remove the host tree from the datastore.
-	_, err = c.client.Delete(ctx, model.IPAMHostKey{Host: hostname}, "")
-	if err != nil {
-		// Return the error unless the resource does not exist.
-		if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
-			log.Errorf("Error removing IPAM host: %v", err)
+	for i := 0; i < ipamEtcdRetries; i++ {
+		// Get the IPAM host.
+		k := model.IPAMHostKey{Host: hostname}
+		kvp, err := c.client.Get(ctx, k, "")
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get IPAM host")
 			return err
 		}
+
+		// Remove the host tree from the datastore.
+		_, err = c.client.Delete(ctx, k, kvp.Revision)
+		if err != nil {
+			if _, ok := err.(cerrors.ErrorResourceUpdateConflict); ok {
+				// We hit a compare-and-delete error - retry.
```

Comment:
This deletion routine feels a bit fishy; it doesn't take a revision so we could delete someone else's concurrent updates.  What happens if an affinity is created between the release above and this retry here?  Does this deletion do a prefix delete (and maybe leave a dangling block)?

Reply:
I think it does take a revision, so if a new affinity is created it should fail the deletion.

But, it will then retry and if it succeeds I think it will leave a dangling block. I think we probably need to pull the releasing of affinities into this retry loop.

So in the loop we:
- Release affinities
- Get IPAMHost
- Compare-and-delete IPAMHost